Add docs for default encoding, discriminator or content type negotiation (#2181)
microsoft/typespec: docs/standard-library/discriminated-types.md

# Discriminated types

> TypeSpec allows you to represent unions and inheritance however when sending types over the wire many languages need a way to discriminate between the various union variants or model hierarchy.

Reviewer: ```suggestion
TypeSpec can express unions and inheritance. However, when sending types over the wire many languages need a way to discriminate between the various union variants or models in an inheritance hierarchy.
```

TypeSpec provide the [`@discriminator` decorator](./built-in-decorators#@discriminator) to be able to help with this pattern.

### Using polymorphism

#### `string` discriminator

```typespec
@discriminator("kind")
model Pet {
  name: string;
  weight?: float32;
}
model Cat extends Pet {
  kind: "cat";
  meow: int32;
}
model Dog extends Pet {
  kind: "dog";
  bark: string;
}
```

#### `enum` discriminator

```typespec
enum PetKind {
  cat,
  dog,
}

@discriminator("kind")
model Pet {
  kind: PetKind;
  name: string;
  weight?: float32;
}
model Cat extends Pet {
  kind: PetKind.cat;
  meow: int32;
}
model Dog extends Pet {
  kind: PetKind.dog;
  bark: string;
}
```

#### Nested discriminator

```tsp
@discriminator("kind")

> model Pet {

Reviewer: Shouldn't `Pet` have `kind: string`, and `Cat` have `breed: string`?

Author: It doesn’t need to but that’s maybe a legacy of not allowing property override. Do you think it should be the recommended way?

  name: string;
  weight?: float32;
}

@discriminator("breed")
model Cat extends Pet {
  kind: "cat";
  meow: int32;
}

@discriminator("breed")

> model Siamese extends Pet {

Reviewer: Maybe this is what you're trying to show but I'm confused why these don't extend `cat`.

Author: Nope it’s a mistake
